File: library/COMN/libCOMN/cpp_comn/istring.cpp
#include "istring.hh"

static const coGUID mthdSetUTF8    = coGUIDMake(s,e,t,u,t,f,8,_,_,_,_,_,_);
static const coGUID mthdUTF8Data   = coGUIDMake(u,t,f,8,d,a,t,a,_,_,_,_,_);
static const coGUID mthdUTF8Length = coGUIDMake(u,t,f,8,l,e,n,g,t,h,_,_,_);

void iString::setUTF8(int64_t data, int32_t length) {
    call(mthdSetUTF8, data, length);
}

int64_t iString::utf8Data() {
    return call(mthdUTF8Data, 0LL);
}

int32_t iString::utf8Length() {
    return (int32_t)call(mthdUTF8Length, 0LL);
}

void iString::setString(const std::string &string) {
    auto data = (int64_t)string.data();
    auto length = (int32_t)string.length();
    setUTF8(data, length);
}

std::string iString::string() {
    auto data = (const char *)utf8Data();
    return data != nullptr ? data : "";
}

iString *iStringCallee::component() {
    return (iString *)_component;
}

int64_t iStringCallee::called(coGUID method, iComponentCalleeParams params) {
    if (method == mthdSetUTF8) {
        auto data = params.a0;
        auto length = (int32_t)params.a1;
        component()->setUTF8(data, length);
        return 0;
    } else if (method == mthdUTF8Data) {
        return component()->utf8Data();
    } else if (method == mthdUTF8Length) {
        return component()->utf8Length();
    } else {
        return 0;
    }
}

File: library/COMN/COMN/IString.m
#import "IString.h"

static const COGUID mthdSetUTF8    = COGUIDMake(s,e,t,u,t,f,8,_,_,_,_,_,_);
static const COGUID mthdUTF8Data   = COGUIDMake(u,t,f,8,d,a,t,a,_,_,_,_,_);
static const COGUID mthdUTF8Length = COGUIDMake(u,t,f,8,l,e,n,g,t,h,_,_,_);

@implementation IString

- (void)setUTF8WithData:(int64_t)data length:(int32_t)length {
    [self call:mthdSetUTF8 raw0:data raw1:length];
}

- (int64_t)utf8Data {
    return [self call:mthdUTF8Data raw0:0];
}

- (int32_t)utf8Length {
    return (int32_t)[self call:mthdUTF8Length raw0:0];
}

@end

@interface IStringCallee ()
@property (nonatomic, readonly) IString *component;
@end

@implementation IStringCallee

- (IString *)component {
    return _component;
}

- (void)called:(COGUID)method params:(IComponentCalleeParams *)params ret:(int64_t *)ret {
    if (method == mthdSetUTF8) {
        int64_t data = params.a0;
        int32_t length = (int32_t)params.a1;
        [self.component setUTF8WithData:data length:length];
    } else if (method == mthdUTF8Data) {
        *ret = [self.component utf8Data];
    } else if (method == mthdUTF8Length) {
        *ret = [self.component utf8Length];
    }
}

@end

@implementation IString (Native)

- (void)setNSString:(NSString *)string {
    if (string.UTF8String != NULL) {
        const char *data = string.UTF8String;
        int32_t length = (int32_t)strlen(data);
        [self setUTF8WithData:(int64_t)data length:length];
    } else {
        [self setUTF8WithData:0 length:0];
    }
}

- (NSString *)NSString {
    const char *cstring = (const char *)self.utf8Data;
    return cstring != NULL ? @(cstring) : nil;
}

@end
